File: projects/engineering.purdue.edu/kak/scriptingwithobjects/swocode/chap14/NoThreadInterference.py
#!/usr/bin/env python

###  NoThreadInterference.py

import threading       

import random                  
ran = random.Random()          

lock = threading.Lock()                                              #(A)

#-------------------------   class DataObject  ------------------------- 
class DataObject( object ):                                          #(B)
    def __init__( self ):                                            #(C)
        self.dataItem1 = 50                                          #(D)
        self.dataItem2 = 50                                          #(E)

    def itemSwap(self):                                              #(F)
        lock.acquire()                                               #(G)
        x = -5 + ran.randint(0,10)                                   #(H)
        self.dataItem1 -= x                                          #(I)
        self.keepBusy()                                              #(J)
        self.dataItem2 += x                                          #(K)
        lock.release()                                               #(L)

    def test(self):                                                  #(M)
        lock.acquire()                                               #(N)
        sum = self.dataItem1 + self.dataItem2                        #(O)
        lock.release()                                               #(P)
        print "Sum: %d" % sum                                        #(Q)

    def keepBusy( self ):                                            #(R)
        for i in range(100000): pass                                 #(S)

#---------------------   class RepeatedSwaps   -------------------------
class RepeatedSwaps( threading.Thread ):                             #(T)
    def __init__( self, dataObject ):                                #(U)
        self.dataObject = dataObject                                 #(V)
        threading.Thread.__init__(self)                              #(W)

    def run( self ):                                                 #(X)
        for i in range (500):                                        #(Y)
            self.dataObject.itemSwap()                               #(Z)
            if i % 100 == 0:                                         #(a)
                self.dataObject.test()                               #(b)

#---------------------------   launch threads -------------------------

d = DataObject()                                                     #(c)

t1 = RepeatedSwaps(d)                                                #(d)
t2 = RepeatedSwaps(d)                                                #(e)
t3 = RepeatedSwaps(d)                                                #(f)

map( lambda x: x.start(), (t1, t2, t3) )                             #(g)

File: projects/demos/engineering.purdue.edu/scriptingwithobjects/swocode/chap14/NoThreadInterference.pl
#!/usr/bin/perl -w

###  NoThreadInterference.pl

use strict;
use threads;
use threads::shared;                       

my @dataObj : shared = (50, 50);                                     #(A)

my $t1 = threads->create( \&repeatedSwaps );                         #(B)
my $t2 = threads->create( \&repeatedSwaps );                         #(C)
my $t3 = threads->create( \&repeatedSwaps );                         #(D)

$t1->join;                                                           #(E)
$t2->join;                                                           #(F)
$t3->join;                                                           #(G)

sub itemSwap {                                                       #(H)
    lock( @dataObj );                                                #(I)
    my $x = int ( -4.999999 + rand(10 ) );                           #(J)
    $dataObj[0] -= $x;                                               #(K)
    keepBusy();                                                      #(L)
    $dataObj[1] += $x;                                               #(M)
}

sub keepBusy {                                                       #(N)
    my $i = 0;                                                       #(O)
    while ( $i++ < 10000 ) {}                                        #(P)
}

sub test {                                                           #(Q)
    lock( @dataObj );                                                #(R)
    my $sum = $dataObj[0] + $dataObj[1];                             #(S)
    print "Sum: $sum\n";                                             #(T)
}

sub repeatedSwaps {                                                  #(U)
    my $j = 0;                                                       #(V)
    while ( $j++ < 500 ) {                                           #(W)
        itemSwap();                                                  #(X)
        test() unless $j % 100;                                      #(Y)
    }
}
